### ISSUE TITLE
Unhandled edge case in _fetch_objects.

### ISSUE THREAD

**Issue by amcgregor** [hmarr/mongoengine#473]
Howdy!

Upgrading from the latest `0.5` version to any version of `0.6` breaks object reloading for me.  I get the following traceback:

```
Traceback (most recent call last):
…
  File "…/src/bachelor/api/user.py", line 428, in add_daily_reward
    self.request.user = self.add_score(self.request.user, reward)
  File "…/src/bachelor/api/user.py", line 387, in add_score
    user.reload()
  File "…/site-packages/mongoengine/document.py", line 299, in reload
    ).first().select_related(max_depth=max_depth)
  File "…/site-packages/mongoengine/document.py", line 287, in select_related
    self._data = DeReference()(self._data, max_depth)
  File "…/site-packages/mongoengine/dereference.py", line 45, in __call__
    self.object_map = self._fetch_objects(doc_type=doc_type)
  File "…/site-packages/mongoengine/dereference.py", line 116, in _fetch_objects
    doc = doc_type._from_son(ref)
AttributeError: 'NoneType' object has no attribute '_from_son'
```

Local variables:

``` python
# File "…/site-packages/mongoengine/document.py", line 299, in reload
{'self': User(4f70bdc1cae33a3af60000aa, "Alice", female), 'max_depth': 1, 'id_field': 'id'}

# File "…/site-packages/mongoengine/document.py", line 287, in select_related
{'self': User(4f70bdc1cae33a3af60000aa, "Alice", female), 'DeReference': <class 'mongoengine.dereference.DeReference'>, 'max_depth': 1}

# File "…/site-packages/mongoengine/dereference.py", line 45, in __call__
{'doc_type': None,
 'instance': None,
 'items': {None: ObjectId('4f70bdc1cae33a3af60000aa'),
           'already_opened': [],
           'answers': {},
           'blackbook': DBRef(u'blackbooks', ObjectId('4f70bdc1cae33a3af60000a9')),
           'facebook': FacebookInfo(100002112804457, http://www.facebook.com/gothalice),
           'gender': u'female',
           'groups': [u'contestants'],
           'id': None,
           'is_dummy': False,
           'is_inactive': False,
           'last_login': datetime.datetime(2012, 3, 26, 19, 4, 35),
           'last_reward': datetime.datetime(2012, 3, 26, 19, 4, 42),
           'name': u'Alice',
           'pending_notices': [],
           'points': PointsInfo(20, 0, 42, 0),
           'random': 0.13644614144844724,
           'tutorials': [u'blind_dates']},
 'max_depth': 1,
 'name': None,
 'self': <mongoengine.dereference.DeReference object at 0x104228b50>}

# File "…/site-packages/mongoengine/dereference.py", line 116, in _fetch_objects
{'col': u'blackbooks',
 'dbref': ObjectId('4f70bdc1cae33a3af60000a9'),
 'dbrefs': [ObjectId('4f70bdc1cae33a3af60000a9')],
 'doc_type': None,
 'keys': [],
 'object_map': {},
 'ref': {u'_id': ObjectId('4f70bdc1cae33a3af60000a9'),
         u'profiles': {},
         u'shared_photos': [],
         u'user_id': ObjectId('4f70bdc1cae33a3af60000aa')},
 'references': <pymongo.cursor.Cursor object at 0x104228dd0>,
 'refs': [ObjectId('4f70bdc1cae33a3af60000a9')],
 'self': <mongoengine.dereference.DeReference object at 0x104228b50>}
```

The problem appears to be the fact that `DeReference.__call__` has no instance (default None), and the code path on lines 35/36 (doc_type defaults to None, no instance, stays None) passes None (the default value) along to `_fetch_objects` on line 45, which then cascades the problem.  The `_fetch_objects` method (starting line 93) defaults doc_type to None, but then can't possibly handle None as a valid value then quite happily attempts to call a method on it on line 116.

Works in 0.5.2, fails in 0.6.0 → 0.6.3, and is rather critical a bug.

Also, unit tests.  -_-;

**Comment by rozza** [comment]
I think the linked commit will fix the issue - but I'm having problems producing a test case to reproduce the actual issue.

**Comment by amcgregor** [comment]
I will test the patch tomorrow^H^H^H^H^H^H^H^H^H today and report back.

**Comment by rozza** [comment]
FYI this branch is on; https://github.com/MongoEngine/mongoengine/tree/473

On Tue, Mar 27, 2012 at 2:30 AM, Alice Zo BevanMcGregor <
reply@reply.github.com

**Comment by amcgregor** [comment]
Now I see the real problem:

```

  File "/Users/amcgregor/Projects/Bachelor/src/bachelor/api/user.py", line 388, in add_score
    user.reload()
  File "/Users/amcgregor/Projects/Bachelor/usr/src/mongoengine/mongoengine/document.py", line 299, in reload
    ).first().select_related(max_depth=max_depth)
  File "/Users/amcgregor/Projects/Bachelor/usr/src/mongoengine/mongoengine/document.py", line 287, in select_related
    self._data = DeReference()(self._data, max_depth)
  File "/Users/amcgregor/Projects/Bachelor/usr/src/mongoengine/mongoengine/dereference.py", line 45, in __call__
    self.object_map = self._fetch_objects(doc_type=doc_type)
  File "/Users/amcgregor/Projects/Bachelor/usr/src/mongoengine/mongoengine/dereference.py", line 118, in _fetch_objects
    doc = get_document(col)._from_son(ref)
  File "/Users/amcgregor/Projects/Bachelor/usr/src/mongoengine/mongoengine/base.py", line 93, in get_document
    """.strip() % name)
NotRegistered: `blackbooks` has not been registered in the document registry.
            Importing the document class automatically registers it, has it
            been imported?
```

My model begins thus:

``` python
class Blackbook(db.Document):
    meta = dict(allow_inheritance=False, collection="blackbooks")
```

And it is being imported.  This keeps getting stranger…

**Comment by rozza** [comment]
Hmm - any further along with this?  Whats in the `base._document_registry` ?

**Comment by amcgregor** [comment]
Unfortunately I haven't been able to proceed with diagnosing this much.  However, here's the `base._document_registry` contents:

``` python
{'Audition': <class 'bachelor.model.show.Audition'>,
 'Auditioner': <class 'bachelor.model.show.Auditioner'>,
 'Blackbook': <class 'bachelor.model.user.Blackbook'>,
 'BlackbookMessage': <class 'bachelor.model.user.BlackbookMessage'>,
 'BlackbookMessageSet': <class 'bachelor.model.user.BlackbookMessageSet'>,
 'BlackbookProfile': <class 'bachelor.model.user.BlackbookProfile'>,
 'CachedFacebookProfile': <class 'bachelor.model.user.CachedFacebookProfile'>,
 'Category': <class 'bachelor.model.show.Category'>,
 'Contestant': <class 'bachelor.model.show.Contestant'>,
 'Crush': <class 'bachelor.model.crush.Crush'>,
 'Dater': <class 'bachelor.model.crush.Dater'>,
 'Event': <class 'bachelor.model.event.Event'>,
 'FacebookInfo': <class 'bachelor.model.user.FacebookInfo'>,
 'Finale': <class 'bachelor.model.show.Finale'>,
 'FutureCrush': <class 'bachelor.model.crush.FutureCrush'>,
 'Notice': <class 'bachelor.model.notice.Notice'>,
 'Notice.CrushNotice': <class 'bachelor.model.notice.CrushNotice'>,
 'Notice.EventNotice': <class 'bachelor.model.notice.EventNotice'>,
 'Notice.ShowNotice': <class 'bachelor.model.notice.ShowNotice'>,
 'PointsInfo': <class 'bachelor.model.user.PointsInfo'>,
 'Question': <class 'bachelor.model.show.Question'>,
 'Quiz': <class 'bachelor.model.show.Quiz'>,
 'QuizBData': <class 'bachelor.model.show.QuizBData'>,
 'QuizCData': <class 'bachelor.model.show.QuizCData'>,
 'RedLight': <class 'bachelor.model.notice.RedLight'>,
 'Show': <class 'bachelor.model.show.Show'>,
 'ShowDump': <class 'bachelor.model.show.ShowDump'>,
 'Situation': <class 'bachelor.model.show.Situation'>,
 'SituationBData': <class 'bachelor.model.show.SituationBData'>,
 'SituationCData': <class 'bachelor.model.show.SituationCData'>,
 'SituationSchema': <class 'bachelor.model.show.SituationSchema'>,
 'Timer': <class 'bachelor.model.show.Timer'>,
 'Topic': <class 'bachelor.model.event.Topic'>,
 'TwelveRoses': <class 'bachelor.model.show.TwelveRoses'>,
 'User': <class 'bachelor.model.user.User'>,
 'UserEvent': <class 'bachelor.model.event.UserEvent'>,
 'UserEventProfile': <class 'bachelor.model.event.UserEventProfile'>}
```

Examining this now I think I can see the problem: the difference between the class name and collection name.  This error happens on a freshly bootstrapped database with data populated using MongoEngine.  :/  (And, as mentioned, doesn't happen on versions prior to 0.6.)

**Comment by erinstargazer** [comment]
Not sure why yet, but it seems to originate from the following change for me: https://github.com/hmarr/mongoengine/commit/a6948771d8

**Comment by kolanos** [comment]
I ran into this bug as well, ac93c3a appears to fix it -- at least for me.

**Comment by rozza** [comment]
kolanos - got a test case?

**Comment by kolanos** [comment]
I haven't come up with a simple test case, but it appears to only occur with collection created by mongoengine <0.6. I don't see it if I drop the collection and re-create it with 0.6.x -- which isn't very practical.

**Comment by kolanos** [comment]
I was mistaken, it is not the result of <0.6 collections/objects. My test was flawed.

**Comment by gregbanks** [comment]
got a test case:

``` python
import sys 

from mongoengine import connect, Document, ReferenceField, StringField

class Foo(Document):
    meta = {'allow_inheritance': False}
    bar = ReferenceField('Bar')


class Bar(Document):
    meta = {'allow_inheritance': False}
    msg = StringField(required=True, default='Blammo!')


def main():
    connect('test_reload')
    Foo.drop_collection()
    Bar.drop_collection()

    bar = Bar()
    bar.save()
    foo = Foo()
    foo.bar = bar 
    foo.save()
    foo.reload()

if __name__ == '__main__':
    sys.exit(main())
```

**Comment by gregbanks** [comment]
Any update on this?

**Comment by rozza** [comment]
Yes - being merged now

**Comment by rozza** [comment]
Fixed in master - will be in 0.6.7

**Comment by gregbanks** [comment]
Rad, thanks!
